### PR TITLE
feat: add dynamic keys for encodeKeyName

### LINKED ISSUE #161: `encodeKeyName` should support dynamic key names
Cf https://github.com/ERC725Alliance/erc725.js/issues/137

Add a new way to encode dynamic keys such as:

```js
myErc725.encodeKeyName('LSP3Profile'); // non dynamic keys -> string
// 0x5ef83ad9559033e6e941db7d7c495acdce616347d28e90c7ce47cbfcfcad3bc5

// multi variables
myErc725.encodeKeyName('AddressPermissions:<string>:<address>', [
  'Permissions',
  '0xa3e6F38477D45727F6e6f853Cdb479b0D60c0aC9',
]);
// 0x4b80742d0000000082ac0000a3e6F38477D45727F6e6f853Cdb479b0D60c0aC9

// single variable
myErc725.encodeKeyName(
  'LSP12IssuedAssetsMap:<address>',
  '0xa3e6F38477D45727F6e6f853Cdb479b0D60c0aC9',
);
// 0x74ac2555c10b934900000000a3e6F38477D45727F6e6f853Cdb479b0D60c0aC9

// single variable in array
myErc725.encodeKeyName('LSP12IssuedAssetsMap:<address>', [
  '0xa3e6F38477D45727F6e6f853Cdb479b0D60c0aC9',
]);
// 0x74ac2555c10b934900000000a3e6F38477D45727F6e6f853Cdb479b0D60c0aC9
```

### Supported variable types:

- https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-2-ERC725YJSONSchema.md#mapping

## Resources:

- <https://docs.soliditylang.org/en/v0.8.14/abi-spec.html#types>

### PR DESCRIPTION
**Pull request by Hugoo**
Closes #161

## TODO:

- [ ] Add `<intM>` support -> EDIT: we won't support it for now.

~~Merge #167 first~~

image

### CODE REVIEW

**Review by frozeman — APPROVED** [review]

File: src/lib/encodeKeyName.ts [hunk 1]
```diff
@@ -1,23 +1,261 @@
...
+
+      if (hexNumber.length <= bytes * 2) {
+        return padLeft(hexNumber, bytes * 2);
+      }
+
+      return hexNumber.slice(-bytes * 2);
+    }
+    case 'int':
+      // TODO:
+      throw new Error('TODO');
```
Comment (Hugoo) [inline src/lib/encodeKeyName.ts:94]: this one is missing

File: docs/classes/ERC725.md [hunk 2]
```diff
@@ -301,9 +301,10 @@ Hashes a key name for use on an [ERC725Y contract](https://github.com/ethereum/E
 
 #### Parameters
 
-| Name      | Type   | Description                      |
-| :-------- | :----- | :------------------------------- |
-| `keyName` | string | The key name you want to encode. |
+| Name                         | Type                           | Description                                  |
+| :--------------------------- | :----------------------------- | :------------------------------------------- |
+| `keyName`                    | string                         | The key name you want to encode.             |
+| `dynamicKeyParts` (optional) | string or <br/> string[&nbsp;] | The variables to use to encode the key name. |
```
Comment (CallumGrindle) [inline docs/classes/ERC725.md:307]:
```suggestion
| `dynamicKeyParts` (optional) | string or <br/> string[&nbsp;] | The variables used to encode the key name. |
```

**Review by CallumGrindle — APPROVED** [review]
Really cool

File: src/lib/encodeKeyName.ts [hunk 3]
```diff
@@ -1,23 +1,261 @@
...
+    );
+  }
+
+  const dynamicKeyPartsArray =
+    typeof dynamicKeyParts === 'string' ? [dynamicKeyParts] : dynamicKeyParts;
+
+  const keyType = guessKeyTypeFromKeyName(name);
+
+  switch (keyType) {
+    case 'Mapping': {
```
[the remark this answers is not in the document]
Comment (Hugoo) [reply src/lib/encodeKeyName.ts]: Good idea thanks, here it is:
https://github.com/ERC725Alliance/erc725.js/pull/168/commits/5f0eb19b4fbffb04b8e8933eea462d3998335ecd

**Review by frozeman — APPROVED** [review]

**Pull request merged**